nexus/src/db/datastore.rs

```diff
@@ -3394,6 +3396,93 @@ impl DataStore {
                 ))
             })
     }
+
+    pub async fn allocate_static_v6_address(
+        &self,
+        allocation_config: &StaticV6AllocationConfig,
+    ) -> Result<Ipv6Addr, Error> {
+        use db::schema::static_v6_address::dsl;
+
+        let allocation_config = allocation_config.clone();
+
+        self.pool()
+            .transaction(move |conn| {
```

Comment:
I'm not sure how important this is, but it'll definitely be feasible to allocate addresses in a single query without a transcation. The [`push_next_available_ip_subquery`](https://github.com/oxidecomputer/omicron/blob/b259f45509bb9a5ba0bb2b656720d1d0472bc6e0/nexus/src/db/subnet_allocation.rs#L721) function is probably a good place to start. One could a left outer join with the table itself, joining on the address being non-NULL, and taking only the first.

For example, this query should work OK:

```
 select 'fd00::'::INET + off as addr from generate_series(0, 10) as off left outer join static_v6_address on (address = 'fd00::'::INET + off) WHERE address is NULL LIMIT 1;
```

That's basically the same thing we have in the linked function, so it might be nice to factor it if possible. That takes the first address where there is no corresponding row already in the `static_v6_address` table. It's effectively the query you have here in the Rust transaction, just in one query. It's still linear in the number of allocated addresses, but requires no transaction or multiple attempts. Hope that helps!

Reply:
Yeah, that's true. Part of the reason we're using IPv6 is that each sled has an entire `/64` to work with. That's...a lot of addresses, so just picking the "next" rather than the lowest unallocated seems OK in the short term. Neither is great in the long term, but the "next after the last allocated address" approach is definitely simpler. The "first fit" one in the linked function makes more sense there, because we're talking about potentially tiny subnets in a customer's VPC Subnet, which can be as small as an IPv4 `/26` I believe. That's only 16 addresses, 6 of which we reserve for ourselves!